Dockerfile

```
@@ -125,16 +239,19 @@ RUN echo "Pip version: ${PIP_VERSION}"
 
 RUN pip install --upgrade pip==${PIP_VERSION}
 
+# We are copying everything with airflow:airflow user:group even if we use root to run the scripts
+# This is fine as root user will be able to use those dirs anyway.
+
 # Airflow sources change frequently but dependency configuration won't change that often
 # We copy setup.py and other files needed to perform setup of dependencies
 # This way cache here will only be invalidated if any of the
 # version/setup configuration change but not when airflow sources change
-COPY --chown=airflow:airflow setup.py /opt/airflow/setup.py
-COPY --chown=airflow:airflow setup.cfg /opt/airflow/setup.cfg
+COPY --chown=airflow:airflow setup.py ${AIRFLOW_HOME}/setup.py
+COPY --chown=airflow:airflow setup.cfg ${AIRFLOW_HOME}/setup.cfg
 
-COPY --chown=airflow:airflow airflow/version.py /opt/airflow/airflow/version.py
-COPY --chown=airflow:airflow airflow/__init__.py /opt/airflow/airflow/__init__.py
-COPY --chown=airflow:airflow airflow/bin/airflow /opt/airflow/airflow/bin/airflow
+COPY --chown=airflow:airflow airflow/version.py ${AIRFLOW_HOME}/airflow/version.py
```

These feel out of place as they wouldn't appear in a airflow home repo. Now the `~airflow` is not the same as `AIRFLOW_HOME` should  we swap all these to

```suggestion
COPY --chown=airflow:airflow airflow/version.py ${AIRFLOW_USER_HOME}/airflow/version.py
```

---

Can you also use:

```
COPY --chown=${AIRFLOW_USER}:${AIRFLOW_USER} ...
```

If not then we probably shouldn't make the user an arg/env as we wouldn't _actually_ be able to change it.

---

Unfortunately not. Current stable version of Docker does not support variables in --chown :(. (At least two months ago). I will check the current version to double-check though and in the future if they add support for it we can change it then still.

The user as arg is still needed. I wanted to keep using airflow as user in the "main" Docker (it is more secure and kind of expected from an official image) but still use root in CI. Running root in CI allows to have local sources mounted directly to the container on both Mac and Linux for development purposes. 

Unfortunately running as root is the only (for now) feasible option for sharing the sources in the local development environment - otherwise you have to map host local user to container airflow which requires some (root required) modifications of docker usermap. Here is a relevant issue described Stack Overflow:
https://stackoverflow.com/questions/49538616/docker-userns-remap-cannot-write-to-mounted-directory/49600083#49600083

So summarizing - I want to use airflow as the owner of folders but then in CI docker I want to use root to run airflow as it allows to have the sources mounted from host to the container easily. I will make some comments in the Dockerfile/scripts to explain that.

---

Gotcha.

This is only an issue on Linux host I think. On Docer-for-mac the volumes go via a remote-fs ("p9" I think now) that maps everything on to the local user on OSX.

---

Yep.

---

@ashb - as for the AIRFLOW_HOME (/home/airflow) vs. AIRFLOW_USER_HOME (/opt/airflow)- this is what we currently have even in the current CI. AIRFLOW_HOME is where the airflow sources are located and where the DAGS and LOGS are eventually placed and where version.py and airflow/bin actually belong I think. 

AIRFLOW_USER_HOME (/home/airflow) is where the home directory of the airflow user with all the .bash_profile scripts etc., and I don't think it's best place for version.py etc.. 

They were two distinct locations even in the original CI scripts. The sources were originally in /app and AIRFLOW_HOME was /home/airflow. It was not possible to have them in one folder because the whole airflow sources directory was mounted as /app folder (so it would overrride all the "home" scripts like .bash_profile etc.:
From: https://github.com/apache/airflow/blob/master/scripts/ci/docker-compose.yml
```
volumes:
      - ../../:/app
```

Hovewer since you raised it - In the current setup we could combine those two. Because in the new Docker-compose configuration I only selectively mount only a subset of sources:

From: https://github.com/PolideaInternal/airflow/blob/simplified-development-workflow/scripts/ci/docker-compose-local.yml
```
    volumes:
      - ../../airflow:/opt/airflow/airflow
      - ../../logs:/opt/airflow/logs
      - ../../dags:/opt/airflow/dags
      - ../../dev:/opt/airflow/dev
      - ../../docs:/opt/airflow/docs
      - ../../scripts:/opt/airflow/scripts
      - ../../tests:/opt/airflow/tests
....
```
 
So we could have AIRFLOW_HOME point to /home/airflow and have /home/airflow/airflow, /home/airflow/dags etc. 

Do you think that it is better approach this way ?

---

I would still suggest not having the airflow source live under a `AIRFLOW_HOME` - that way that AIRFLOW_HOME looks more "typical" of what an end user would have.

---

Hey @ashb - so just to verify If understand correctly - we want to separate sources (we need them to run tests) from AIRFLOW_HOME (where logs and dags are stored):
* HOME -> the usual HOME of user (either /home/airflow or /root in CI)
* AIRFLOW_HOME -> ${HOME}/airflow ( /dags and /logs are subdirectories )
* AIRFLOW_SOURCES (i propose to name it this way) -> /opt/airflow - this is where mounted sources of airflow should be.

Is my understanding correct ?

---

Yes, exactly that, and the new name is much env var name is clearer too. I like.